File: src/main/java/cn/labzen/web/ap/evaluate/annotation/RequestMappingEvaluator.java
package cn.labzen.web.ap.evaluate.annotation;

import cn.labzen.web.ap.config.Config;
import cn.labzen.web.ap.internal.Utils;
import cn.labzen.web.ap.internal.context.AnnotationProcessorContext;
import cn.labzen.web.ap.internal.element.ElementAnnotation;
import cn.labzen.web.ap.internal.element.ElementMethod;
import cn.labzen.web.ap.suggestion.AppendSuggestion;
import cn.labzen.web.ap.suggestion.ReplaceSuggestion;
import cn.labzen.web.ap.suggestion.Suggestion;
import cn.labzen.web.defination.APIVersionCarrier;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;

import java.util.Collections;
import java.util.List;
import java.util.Map;

import static cn.labzen.web.ap.definition.TypeNames.ANNOTATION_API_VERSION;
import static cn.labzen.web.ap.definition.TypeNames.ANNOTATION_SPRING_REQUEST_MAPPING;

public final class RequestMappingEvaluator implements MethodAnnotationErasableEvaluator {

  private TypeName apiVersionType;

  @Override
  public boolean support(TypeName type) {
    return type instanceof ClassName cn && Utils.isRequestMappingAnnotation(cn);
  }

  @Override
  public void init(AnnotationProcessorContext context) {
    apiVersionType = TypeName.get(context.elements().getTypeElement(ANNOTATION_API_VERSION).asType());
  }

  @Override
  public List<? extends Suggestion> evaluate(Config config, TypeName type, Map<String, Object> members) {
    APIVersionCarrier carrier = config.apiVersionCarrier();
    if (carrier == APIVersionCarrier.DISABLE) {
      return Collections.emptyList();
    }

    String version = config.apiVersionPrefix() + config.apiVersionBased();
    var annotation = switch (carrier) {
      case URI -> versionByURI(version);
      case HEADER -> versionByHeader(config, version, type);
      case PARAMETER -> versionByParameter(config, version, type);
      default -> throw new IllegalStateException("never happen");
    };

    return List.of(annotation);
  }

  /**
   * 通过 URI 版本控制
   */
  private Suggestion versionByURI(String version) {
    ElementAnnotation annotation = new ElementAnnotation(apiVersionType, Map.of("value", version));
    return new AppendSuggestion(annotation, ElementMethod.class);
  }

  /**
   * 通过 Header Accept 版本控制
   */
  private Suggestion versionByHeader(Config config, String version, TypeName type) {
    String headerVersion = "application/vnd." + config.apiVersionHeaderVND() + "." + version + "+json";
    ElementAnnotation annotation = new ElementAnnotation(type, Map.of("produces", new String[]{headerVersion}));
    return new ReplaceSuggestion(ANNOTATION_SPRING_REQUEST_MAPPING, annotation);
  }

  /**
   * 通过请求参数控制版本
   */
  private Suggestion versionByParameter(Config config, String version, TypeName type) {
    String paramVersion = config.apiVersionParameterName() + "=" + version;
    ElementAnnotation annotation = new ElementAnnotation(type, Map.of("params", new String[]{paramVersion}));
    return new ReplaceSuggestion(ANNOTATION_SPRING_REQUEST_MAPPING, annotation);
  }
}

File: src/main/kotlin/cn/labzen/web/ap/evaluate/annotation/RequestMappingEvaluator.kt
package cn.labzen.web.ap.evaluate.annotation

import cn.labzen.web.ap.config.Config
import cn.labzen.web.ap.internal.Utils
import cn.labzen.web.ap.internal.element.ElementAnnotation
import cn.labzen.web.ap.internal.element.ElementMethod
import cn.labzen.web.ap.suggestion.Suggestion
import cn.labzen.web.ap.suggestion.impl.AppendSuggestion
import cn.labzen.web.ap.suggestion.impl.ReplaceSuggestion
import cn.labzen.web.defination.APIVersionCarrier.*
import cn.labzen.web.runtime.annotation.APIVersion
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.TypeName
import org.springframework.web.bind.annotation.RequestMapping

class RequestMappingEvaluator : MethodErasableAnnotationEvaluator {

  override fun support(type: TypeName): Boolean =
    type is ClassName && Utils.isRequestMappingAnnotation(type)

  override fun evaluate(config: Config, type: TypeName, members: Map<String, Any?>): List<Suggestion> {
    if (config.apiVersionCarrier() == DISABLE) {
      return emptyList()
    }

    val version = config.apiVersionPrefix() + config.apiVersionBased()
    val annotationSuggestion = when (config.apiVersionCarrier()) {
      URI -> versionByURI(version)
      HEADER -> versionByHeader(config, version, type)
      PARAMETER -> versionByParameter(config, version, type)
      else -> throw IllegalStateException("never happen")
    }

    return listOf(annotationSuggestion)
  }

  /**
   * 通过 URI 版本控制
   */
  private fun versionByURI(version: String): Suggestion {
    val annotation = ElementAnnotation(TypeName.get(APIVersion::class.java), mutableMapOf("value" to version))
    return AppendSuggestion(annotation, ElementMethod::class.java)
  }

  /**
   * 通过 Header Accept 版本控制
   */
  private fun versionByHeader(config: Config, version: String, clazz: TypeName): Suggestion {
    val headerVersion = "application/vnd.${config.apiVersionHeaderVND()}.$version+json"
    val annotation = ElementAnnotation(clazz, mutableMapOf("produces" to arrayOf(headerVersion)))
    return ReplaceSuggestion(SUPPORTED_NAME, annotation)
  }

  /**
   * 通过请求参数控制版本
   */
  private fun versionByParameter(config: Config, version: String, clazz: TypeName): Suggestion {
    val paramVersion = "${config.apiVersionParameterName()}=$version"
    val annotation = ElementAnnotation(clazz, mutableMapOf("params" to arrayOf(paramVersion)))
    return ReplaceSuggestion(SUPPORTED_NAME, annotation)
  }

  companion object {
    private val SUPPORTED = TypeName.get(RequestMapping::class.java)
    private val SUPPORTED_NAME = Utils.getSimpleName(SUPPORTED)
  }
}